### ISSUE TITLE
Avoid unnecessary linking to libboost_unit_test_framework.so 

### ISSUE THREAD

**Issue by springmeyer** [Project-OSRM/osrm-backend#2065]
The command line tools all link libboost_unit_test_framework even though it is not used by osrm-extract, osrm-contract, etc. This is harmless for local development. But when distributing binaries of the command line tools it becomes problematic bloat: unless this unnecessary shared library is also distributed the tools will fail to run.

So, overall it would be ideal to ensure each tool only links the libraries it actually needs. This might also have the added benefit that linking will be faster.

**Comment by daniel-j-h** [comment]
> But when distributing binaries of the command line tools it becomes problematic bloat: unless this unnecessary shared library is also distributed the tools will fail to run.

Really? I can not see `libboost_unit_test_framework.so` in the dynamic sections for `osrm-extract` or even `libosrm.so`:

```
daniel@x1c /t/o/build> readelf -d osrm-extract|ag needed
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libosrm_extract.so]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_filesystem.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_program_options.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_system.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libtbb.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
```

```
daniel@x1c /t/o/build> readelf -d libosrm.so|ag needed
 0x0000000000000001 (NEEDED)             Shared library: [libboost_filesystem.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_system.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_thread.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

Same for the statically linked `libosrm.a`:

```
daniel@x1c /t/o/build> gcc-nm -C libosrm.a|ag unit_test|wc -l
0
```

It's in the `extractor-tests` though:

```
daniel@x1c /t/o/build> readelf -d extractor-tests|ag needed
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libbz2.so.1.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_filesystem.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_regex.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_system.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libboost_unit_test_framework.so.1.59.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libexpat.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libluabind.so]
 0x0000000000000001 (NEEDED)             Shared library: [liblua5.2.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstxxl.so.1.4.99]
 0x0000000000000001 (NEEDED)             Shared library: [libtbb.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgomp.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```

Somewhat related: we even discard unused symbols by first artificially bloating ELF sections so we're able to discard them later ([ref](https://github.com/Project-OSRM/osrm-backend/blob/0c5899b2a258f22e1c9438df0e98deed60fee1af/CMakeLists.txt#L192-L195)).

You're right, it's not a clean approach to link in everything and rely on the the linker to throw away unused dependencies. I'm still curious: does it really break things?

**Comment by TheMarex** [comment]
@daniel-j-h is there an easy way to factor out the boost module includes?

**Comment by daniel-j-h** [comment]
> boost module includes?

No idea what you mean, but assuming the Boost.UnitTest includes are only needed in the tests this is how to not link the `unit_test_framework` in except for the tests in the first place:

- [ ] splitt off `unit_test_framework` from other Boost modules we require in [CMakeLists.txt](https://github.com/Project-OSRM/osrm-backend/blob/adc3bacea960b7bdd8d01f24c21454a9c0f1fff5/CMakeLists.txt#L51)
- [ ] still do a FindPackage on Boost and [require](https://github.com/Project-OSRM/osrm-backend/blob/adc3bacea960b7bdd8d01f24c21454a9c0f1fff5/CMakeLists.txt#L236) `unit_test_framework`
- [ ] using `target_link_libraries` do not link against `unit_test_framework` except for the tests

**Comment by springmeyer** [comment]
@daniel-j-h I was seeing this on osx and assumed the same behavior on linux. You are correct that linux `readelf` or `ldd` output shows that libraries that are linked unnecessarily are then automatically dropped (just checked). But this is not the case on OS X (`otool -L`):

```
$ otool -L build/osrm-extract 
build/osrm-extract:
	/Users/dane/.homebrew/opt/boost/lib/libboost_date_time-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_filesystem-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_iostreams-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_program_options-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_regex-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_system-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_thread-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_unit_test_framework-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/Users/dane/.homebrew/opt/luabind/lib/libluabind.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/lua51/lib/liblua.5.1.dylib (compatibility version 5.1.0, current version 5.1.5)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/Users/dane/.homebrew/lib/libtbb.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/lib/libtbbmalloc.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

**Comment by springmeyer** [comment]
>  You are correct that linux readelf or ldd output shows that libraries that are linked unnecessarily are then automatically dropped (just checked).

Sorry, I was wrong. The `ldd` output on linux does in fact show that `osrm-extract`, for example, depends on `libboost_unit_test_framework.so` and therefore will be broken if the dynamic library is not available. The reason I mis-interpreted this is that previously I was looking at `ldd` output when statically linking all the boost libraries. 

```
$ ldd osrm-extract 
	linux-vdso.so.1 =>  (0x00007fff8b98d000)
	libboost_date_time.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_date_time.so.1.55.0 (0x00007fe332839000)
	libboost_filesystem.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.55.0 (0x00007fe332623000)
	libboost_iostreams.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_iostreams.so.1.55.0 (0x00007fe33240a000)
	libboost_program_options.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_program_options.so.1.55.0 (0x00007fe33219a000)
	libboost_regex.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_regex.so.1.55.0 (0x00007fe331e96000)
	libboost_system.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.55.0 (0x00007fe331c91000)
	libboost_thread.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.55.0 (0x00007fe331a7a000)
	libboost_unit_test_framework.so.1.55.0 => /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.55.0 (0x00007fe3317d7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe3315b9000)
```

**Comment by daniel-j-h** [comment]
@springmeyer I started a modular-cmake branch over at https://github.com/Project-OSRM/osrm-backend/pull/2073 which is long overdue anyway. Can you check and confirm that this resolves the issue?

**Comment by springmeyer** [comment]
Excellent, thank you @daniel-j-h! Looking solved (no `libboost_unit_test_framework` in `otool` output)!

```
$ otool -L build/osrm-extract 
build/osrm-extract:
	/Users/dane/.homebrew/opt/boost/lib/libboost_date_time-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_filesystem-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_iostreams-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_program_options-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_regex-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_system-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/boost/lib/libboost_thread-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/Users/dane/.homebrew/opt/luabind/lib/libluabind.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/opt/lua51/lib/liblua.5.1.dylib (compatibility version 5.1.0, current version 5.1.5)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/Users/dane/.homebrew/lib/libtbb.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/dane/.homebrew/lib/libtbbmalloc.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

**Comment by springmeyer** [comment]
Closing, confirmed this is solved by #2227
